lib/msf/base/sessions/ssh_command_shell_bind.rb

```diff
@@ -187,28 +258,84 @@ def create(params)
       # Notify handlers before we create the socket
       notify_before_socket_create(self, params)
 
-      mutex = Mutex.new
-      condition = ConditionVariable.new
-      ssh_channel = msf_channel = nil
-
       if params.proto == 'tcp'
         if params.server
-          raise ::Rex::BindFailed.new(params.localhost, params.localport, reason: 'TCP server sockets are not supported by SSH sessions.')
-        end
-
-        ssh_channel = @ssh_connection.open_channel('direct-tcpip', :string, params.peerhost, :long, params.peerport, :string, params.localhost, :long, params.localport) do |new_channel|
-          dlog("new direct-tcpip channel opened to #{Rex::Socket.is_ipv6?(params.peerhost) ? '[' + params.peerhost + ']' : params.peerhost}:#{params.peerport}")
-          msf_channel = TcpClientChannel.new(self, @channel_ticker += 1, new_channel, params)
-          mutex.synchronize do
-            condition.signal
-          end
+          sock = create_server_channel(params)
+        else
+          sock = create_client_channel(params)
```

Comment:
So these methods are called create channel, and it looks like `#create_server_channel` returns a channel, while `#create_client_channel` returns a socket. In both cases however the return value is assigned to sock implying it's a socket and then treated as such. Should these both be returning channels, or should these both be returning sockets? In this case it seems incorrect that the server channel is being passed to `notify_socket_created`.

Reply:
So for this, I basically followed a similar implementation to Meterpreter's "TcpServerChannel" (https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_server_channel.rb), and the naming conventions I found there. That one refers to a "Server channel", which to my understanding, implements the socket interface by doing the following:
- Waits for TLVs saying "Please open a ClientChannel", and then create one
- Wait for a user of the class to call `accept` (in a socket-like way), at which point, return an available ClientChannel

I've basically taken the same approach for the SSH implementation.

If I've misunderstood how it all fits together, let me know; but otherwise I suspect maybe the naming conventions are working against making this all easy to follow? Certainly took me a while to deduce that TcpServerChannels were pretending to be a socket.

Reply:
I see what you're saying, you're replicating [this pattern here](https://github.com/rapid7/metasploit-framework/blob/d1d680ed024a09f9c5e36b458a550447772df7ef/lib/rex/post/meterpreter/extensions/stdapi/net/socket.rb#L105-L136). I'd never noticed that before but you're right. I appreciate your attention to detail! :+1: